config/auth_azure_cli.go

```diff
@@ -50,11 +54,32 @@ func (c AzureCliCredentials) Configure(ctx context.Context, cfg *Config) (func(*
 		return nil, fmt.Errorf("resolve host: %w", err)
 	}
 	logger.Infof(ctx, "Using Azure CLI authentication with AAD tokens")
-	return refreshableVisitor(&ts), nil
+
+	// There are three scenarios:
+	//
+	//  1. The user has logged in with the Azure CLI as a user and has access to the service management endpoint.
+	//  2. The user has logged in with the Azure CLI as a user and does not have access to the service management endpoint.
+	//  3. The user has logged in with the Azure CLI as a service principal, and must have access to the service management
+	//     endpoint to authenticate.
+	//
+	// If the user can't access the service management endpoint, we assume they are in case 2 and do not pass the service
+	// management token. Otherwise, we always pass the service management token.
+	var managementTs oauth2.TokenSource
+	env, err := cfg.GetAzureEnvironment()
+	if err != nil {
+		return nil, err
+	}
+	managementTs = &azureCliTokenSource{env.ServiceManagementEndpoint, ""}
+	_, err = managementTs.Token()
+	if err != nil {
+		managementTs = nil
```

Comment:
Debug log with the error would be nice. Will make all the difference debugging this not working.

Reply:
Separate comment, why call `.Token()` in the first place if we ignore the error in the visitor if we call it there? I suspect this is to avoid calling it repeatedly if we don't have one anyway. If so, please add comment.

Reply:
And, if we do already pull this token eagerly, we should not tolerate errors on the secondary token source (the final arg to the visitor function).

Reply:
I kind of blended two changes together. Originally, I had a version that would check for the management token when making requests and tolerating failures there. Partway through, I changed it to a version that checked for the management token during Configure() and skips even passing it to the visitor factory. You're right, we don't need to tolerate errors (and we can do one better: just call the refreshableVisitor method instead).